src/analysis/processing/qgsalgorithmb3dmtogltf.cpp

```diff
@@ -0,0 +1,128 @@
+
+QString QgsB3DMToGltfAlgorithm::name() const
+{
+  return QStringLiteral( "b3dmtogltf" );
+}
+
+QString QgsB3DMToGltfAlgorithm::displayName() const
+{
+  return QObject::tr( "Convert B3DM to GLTF" );
+}
+
+QStringList QgsB3DMToGltfAlgorithm::tags() const
+{
+  return QObject::tr( "3d,tiles,cesium" ).split( ',' );
+}
+
+QString QgsB3DMToGltfAlgorithm::group() const
+{
+  return QObject::tr( "3D Tiles" );
+}
+
+QString QgsB3DMToGltfAlgorithm::groupId() const
+{
+  return QStringLiteral( "3dtiles" );
+}
+
+QString QgsB3DMToGltfAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Converts files from the legacy B3DM format to GLTF." );
+}
+
+QgsB3DMToGltfAlgorithm *QgsB3DMToGltfAlgorithm::createInstance() const
+{
+  return new QgsB3DMToGltfAlgorithm();
+}
+
+void QgsB3DMToGltfAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFile( QStringLiteral( "INPUT" ), QObject::tr( "Input B3DM" ), QgsProcessingParameterFile::File,
+                QStringLiteral( "b3dm" ), QVariant(), false, QStringLiteral( "B3DM (*.b3dm *.B3DM)" ) ) );
+
+  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Output GLTF file" ), QStringLiteral( "GLTF (*.gltf *.GLTF)" ) ) );
+}
+
+QVariantMap QgsB3DMToGltfAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString path = parameterAsFile( parameters, QStringLiteral( "INPUT" ), context );
+  const QString outputPath = parameterAsFile( parameters, QStringLiteral( "OUTPUT" ), context );
+
+  if ( !QFile::exists( path ) )
+    throw QgsProcessingException( QObject::tr( "Could not load source file %1." ).arg( path ) );
+
+  QFile f( path );
+  QByteArray b3dmContent;
+  if ( f.open( QIODevice::ReadOnly ) )
+  {
+    b3dmContent = f.readAll();
+  }
+  else
+  {
+    throw QgsProcessingException( QObject::tr( "Could not load source file %1." ).arg( path ) );
+  }
+
+  const QByteArray gltfContent = QgsCesiumUtils::extractGltfFromB3dm( b3dmContent );
+
+  // load gltf and then rewrite -- this allows us to both validate the B3DM GLTF content, and
+  // also gives the opportunity to "upgrade" B3DM specific options (like CESIUM_RTC) to standard
+  // GLTF extensions
+  tinygltf::Model model;
+  QString errors;
+  QString warnings;
+  if ( !QgsGltfUtils::loadGltfModel( gltfContent, model, &errors, &warnings ) )
+  {
+    throw QgsProcessingException( QObject::tr( "Error loading B3DM model: %1" ).arg( errors ) );
+  }
+  if ( !warnings.isEmpty() )
+  {
+    feedback->pushWarning( warnings );
+  }
+
+  const QByteArray outputFile = QFile::encodeName( outputPath );
+  std::ofstream of( outputFile.constData(), std::ios::binary | std::ios::trunc );
+  if ( !of )
+    throw QgsProcessingException( QObject::tr( "Could not create destination file %1." ).arg( outputPath ) );
+
+  tinygltf::TinyGLTF writer;
+  if ( !writer.WriteGltfSceneToStream( &model, of ) )
```

Comment:
How is this going to handle textures? As we're not using STB_image_write, I assume they would be left out?

Reply:
On the files I tested, it worked ok :shrug: I assume it's just copying the raw binary data across unchanged, so doesn't need to decode/recode the images.

Reply:
Ah ok, cool!